### PR TITLE
[FSDP][optim_state_dict] Fuse allgather for optim_state_dict when use_orig_params is True

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108434
* __->__ #108298

The original implementation of `_gather_orig_param_state` is naive. It performs one allgather_object and two allgather (if the optimizer is Adam) per FQN. This can be slow and make `_optim_state_dict` become bottleneck.

This PR rewrite the implementation and fuse all the `allgather_object`s into one. As for `allgather`, it is fused based on the information of FlatParameters. So there will be 2N `allgather` where N is the number of FlatParameter and 2 is due to Adam having 2 states per FQN.

One experiment on 8GPU A100 shows that the execution of the gathering is improved to 0.3 seconds from 3 seconds.

Differential Revision: [D48835138](https://our.internmc.facebook.com/intern/diff/D48835138/)

### CODE REVIEW

**Comment by fegin** [comment]
@awgu Can you review if the implementation of concat with aligned padding is correct?

File: test/distributed/fsdp/test_fsdp_optim_state.py [hunk 1]
```diff
@@ -142,8 +142,8 @@ def forward(self, x):
 class NestedModel(torch.nn.Module):
     def __init__(self) -> None:
         super().__init__()
-        self.block0 = BlockB(5, 7)
-        self.block1 = BlockB(7, 7)
+        self.block0 = BlockB(5, 3)
```
Comment (wz337) [inline test/distributed/fsdp/test_fsdp_optim_state.py:145]: Noob question: what's the difference between the two?

Comment (fegin) [reply test/distributed/fsdp/test_fsdp_optim_state.py:145]: With this new setting, we were able to reproduce the align padding issue.

Comment (awgu) [reply test/distributed/fsdp/test_fsdp_optim_state.py:145]: This does make me wonder if it would be worth utility to visualize/print the padding.

File: torch/distributed/fsdp/_optim_utils.py [hunk 2]
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
     return isinstance(key, str)
 
 
+@dataclass
+class StateInfo:
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1280]: nit: Do you think it is worth documenting that keys are state names, e.g. `"exp_avg"`?

File: torch/distributed/fsdp/_optim_utils.py [hunk 3]
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
...
+def _convert_state_with_orig_params(
+    all_optim_state_keys: List[_OptimStateKey],
+    optim_state_key_to_param_key: Dict[_OptimStateKey, Union[int, str]],
+    fqn_to_fsdp_param_info: Dict[str, FSDPParamInfo],
+    optim_state_dict: Dict[Union[str, int], Any],
+    to_save: bool,
+    shard_state: bool,
+) -> Dict[str, Any]:
+    fsdp_osd_state: Dict[str, Any] = {}
+    all_states: Dict[int, Dict[str, Any]] = {}
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1606]: Note to self: The key is only used to de-dup against `FSDPParamInfo` objects. One `FSDPParamInfo` object corresponds to one `FlatParamHandle`/`FlatParameter`.

File: torch/distributed/fsdp/_optim_utils.py [hunk 4]
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
...
+    for optim_state_key in all_optim_state_keys:
+        param_key: Union[str, int, None] = optim_state_key_to_param_key.get(
+            optim_state_key, None
+        )
+
+        if param_key is None and not optim_state_key.is_fsdp_managed:
+            continue
+
+        if optim_state_key.is_fsdp_managed:
+            fqn = optim_state_key.unflat_param_names[0]
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1618]: Note to self: For `use_orig_params=True` path, it seems that `unflat_param_names` is always a singleton tuple.

File: torch/distributed/fsdp/_optim_utils.py [hunk 5]
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
...
+                    )
+                elif begin < padding_begin <= padding_end < end:
+                    # This is an align padding that is completely in the shard.
+                    padding_len = numel
+                else:
+                    padding_len = 0
+                if padding_len:
+                    local_buffers.append(empty_func(padding_len))
+            else:
+                if buffers[buffer_idx] is not None:
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1516]: When is this `False`?

File: torch/distributed/fsdp/_optim_utils.py [hunk 6]
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
...
+def _unflatten_orig_param_states(
+    fsdp_param_info: FSDPParamInfo,
+    output_states: Dict[str, Dict[str, Any]],
+    state_name: str,
+    shard_state: bool,
+    to_save: bool,
+) -> None:
+    """
+    Given a output state dict, ``output_states``, which the keys are FQNs to the
+    original parameters (not FlatParameters nor parmeter ID), and the values
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1403]: nit: 
```suggestion
    original parameters (not FlatParameters nor parameter ID), and the values
```

File: torch/distributed/fsdp/_optim_utils.py [hunk 7]
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
...
+            f"shard_numel_padded={shard_numel_padded}, "
+            f"_sharded_size.numel={flat_param._sharded_size.numel()}, "
+            f"_numels_with_padding={flat_param._numels_with_padding}, "
+            f"begin={begin}, end={end},"
+        )
+        if shard_numel_padded > 0:
+            # Add right-handed padding.
+            local_buffers.append(empty_func(shard_numel_padded))
+        local_shard = torch.cat(local_buffers)
+        assert local_shard.numel() * fsdp_state.world_size == gathered_tensor.numel()
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1537]: nit: It looks like if this `assert` fails, then there is a bug inside `flat_param.py`'s metadata.
For completeness, should we add an assert message with the relevant info?

File: torch/distributed/fsdp/_optim_utils.py [hunk 8]
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
...
+            dist.all_gather_into_tensor(
+                gathered_tensor, local_shard, group=fsdp_state.process_group
+            )
+            # Synchronize can be slow but this will be easier for us to debug.
+            torch.cuda.synchronize()
+
+        unpadded_tensor = gathered_tensor[: flat_param._unpadded_unsharded_size.numel()]
+        flat_param_handle = fsdp_param_info.handle
+        orig_states = flat_param_handle._get_unflat_views_aligned(unpadded_tensor)
+        assert len(orig_states) == len(fsdp_param_info.param_indices)
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1549]: nit: Similarly, maybe just include the `assert` message in case?

File: torch/distributed/fsdp/_optim_utils.py [hunk 9]
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
...
+    fsdp_param_info: FSDPParamInfo,
+    input_states: Dict[str, Any],
+    shard_state: bool,
+    to_save: bool,
+) -> Dict[str, Any]:
+    """
+    Given a optimizer state dict, ``input_states``, which the keys are FQNs to the
+    original parameters (not FlatParameters nor parmeter ID), gather all the
+    states and unflatten them to the original dimensions. Note that all the
+    params refered by the ``input_states`` must be manageded by FSDP.
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1575]: nit: 
```suggestion
    params referred by the ``input_states`` must be managed by FSDP.
```

File: torch/distributed/fsdp/_optim_utils.py [hunk 10]
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
...
+        ):
+            if is_padding:
+                padding_begin, padding_end = pos, pos + numel - 1
+                if padding_begin <= begin <= padding_end:
+                    # This is an align padding right before the first non-None
+                    # buffer in the shard. The shard includes parts of the
+                    # align padding.
+                    padding_len = (
+                        padding_end - begin + 1
+                        if end >= padding_end
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1496]: Note to self: `end < padding_end` means that this rank's entire sharded flat parameter is padding since then `padding_begin <= begin < end < padding_end`.

File: torch/distributed/fsdp/_optim_utils.py [hunk 11]
```diff
@@ -1272,6 +1276,436 @@ def _is_named_optimizer(optim_state_dict: Dict[str, Any]) -> bool:
...
+                        else end - begin + 1
+                    )
+                elif padding_begin <= end <= padding_end:
+                    # This is an align padding right after the last non-None
+                    # buffer in the shard. The shard includes parts of the
+                    # align padding.
+                    padding_len = (
+                        end - padding_begin + 1
+                        if begin <= padding_begin
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:1505]: Note to self: Similarly, if `begin > padding_begin`, then `padding_begin < begin < end <= padding_end`, and the rank's entire sharded flat parameter is padding.

**Review by awgu — APPROVED** [review]
The padding logic looks good to me!

File: torch/distributed/fsdp/_optim_utils.py (hunk 5 above)
Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py:1516]: When the rank does not own the buffer. Add a comment to explain.

File: torch/distributed/fsdp/_optim_utils.py (hunk 7 above)
Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py:1537]: It may not be a bug of FlatParameter's metadata but some inconsistency of the reconstruction logic here. Maybe some  edge cases. Added the message.

File: test/distributed/fsdp/test_fsdp_optim_state.py (hunk 1 above)
Comment (fegin) [reply test/distributed/fsdp/test_fsdp_optim_state.py:145]: Yup, since we are going to maintain FSDP FlatParameter version for a while, having simple utility to print the padding is going to be useful. We can make it a ramp-up task.
